Sources/PT.PM.AntlrUtils/AntlrListenerConverter.cs

```diff
@@ -0,0 +1,255 @@
+using System.Collections.Generic;
+using Antlr4.Runtime;
+using Antlr4.Runtime.Tree;
+using PT.PM.Common;
+using PT.PM.Common.Files;
+using PT.PM.Common.Nodes;
+using PT.PM.Common.Nodes.Collections;
+
+namespace PT.PM.AntlrUtils
+{
+    /// <summary>
+    /// Generates UST node at each `ExitEveryRule` method
+    /// If needed, use concrete Listener with overriden methods for certain contexts
+    /// </summary>
+    public abstract class AntlrListenerConverter : IParseTreeListener, ILoggable
+    {
+        private static readonly char[] trimChar = {'\''};
+
+        private int childNodesIndex = -1;
+        private readonly List<Usts> childNodes = new List<Usts>();
+        private readonly AntlrParserConverter antlrParserConverter;
+        private readonly AntlrConvertHelper convertHelper;
+        protected readonly RootUst root;
+
+        public ILogger Logger { get; set; } = DummyLogger.Instance;
+
+        public bool IsParseTreeExisting { get; set; }
+
+        protected AntlrListenerConverter(TextFile sourceFile, AntlrParserConverter antlrParser)
+        {
+            root = new RootUst(sourceFile, antlrParser.Language);
+            antlrParserConverter = antlrParser;
+            convertHelper = new AntlrConvertHelper(root, antlrParser);
+            PushNew(null);
+        }
+
+        /// <summary>
+        /// Calls at the end of conversion process.
+        /// </summary>
+        public RootUst Complete()
+        {
+            ConvertChildren();
+            root.Node = GetChildren()[0];
+            PopChildren();
+            Clear();
+            root.FillAscendants(); // TODO: should be filled during conversion
```

Comment:
Fair remark. Actually, this method fills `Parent` and `Root` properties for all descendants and self. Maybe rename to `FillParentAndRootForDescendatnsAndSelf`?

Reply:
I'd say `FillParentAndRootForSelfAndDescendatns`. :smiley:

Reply:
I prefer my variant because of Roslyn team use the same order: https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.operations.operationextensions.descendantsandself?view=roslyn-dotnet